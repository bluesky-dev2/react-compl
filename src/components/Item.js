import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import styles from './styles';

class Item extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    className: PropTypes.string,
    style: PropTypes.object,
    nativeEvent: PropTypes.object,
    disabled: PropTypes.oneOfType([PropTypes.bool, PropTypes.func]),
    onClick: PropTypes.func,
    data: PropTypes.any,
    dataFromProvider: PropTypes.any,
    refsFromProvider: PropTypes.oneOfType([
      PropTypes.object,
      PropTypes.arrayOf(PropTypes.object)
    ])
  };

  static defaultProps = {
    className: null,
    style: {},
    disabled: false,
    onClick: () => {},
    nativeEvent: {},
    data: null,
    refsFromProvider: [],
    dataFromProvider: null
  };

  disabled = false;

  handleClick = e => {
    this.disabled
      ? e.stopPropagation()
      : this.props.onClick({
          event: this.props.nativeEvent,
          ref: this.props.refsFromProvider,
          data: this.props.data,
          dataFromProvider: this.props.dataFromProvider
        });
  };

  render() {
    const {
      className,
      disabled,
      style,
      children,
      data,
      refsFromProvider,
      dataFromProvider,
      nativeEvent
    } = this.props;
    
    this.disabled =
      typeof disabled === 'function'
        ? disabled({
            event: nativeEvent,
            ref: refsFromProvider,
            data: data,
            dataFromProvider: dataFromProvider
          })
        : disabled;

    const cssClasses = cx(styles.item, className, {
      [`${styles.itemDisabled}`]: this.disabled
    });

    return (
      <div
        className={cssClasses}
        style={style}
        onClick={this.handleClick}
        role="presentation"
      >
        <div className={styles.itemContent}>{children}</div>
      </div>
    );
  }
}

export default Item;
