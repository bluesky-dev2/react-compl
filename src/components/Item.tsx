import React, { ReactNode } from 'react';
import cx from 'clsx';

import {
  ItemParams,
  InternalProps,
  BooleanPredicate,
  HandlerParamsEvent,
} from '../types';
import { useRefTrackerContext } from './RefTrackerProvider';
import { NOOP, STYLE } from '../constants';
import { getPredicateValue } from './utils';

export interface ItemProps
  extends InternalProps,
    Omit<React.HTMLAttributes<HTMLElement>, 'hidden' | 'disabled' | 'onClick'> {
  /**
   * Any valid node that can be rendered
   */
  children: ReactNode;

  /**
   * Passed to the `Item` onClick callback. Accessible via `data`
   */
  data?: any;

  /**
   * Disable `Item`. If a function is used, a boolean must be returned
   *
   * @param props The props passed when you called `show(e, {props: yourProps})`
   * @param data The data defined on the `Item`
   * @param triggerEvent The event that triggered the context menu
   *
   *
   * ```
   * function isItemDisabled({ triggerEvent, props, data }: PredicateParams<type of props, type of data>): boolean
   * <Item disabled={isItemDisabled} data={data}>content</Item>
   * ```
   */
  disabled?: BooleanPredicate;

  /**
   * Hide the `Item`. If a function is used, a boolean must be returned
   *
   * @param props The props passed when you called `show(e, {props: yourProps})`
   * @param data The data defined on the `Item`
   * @param triggerEvent The event that triggered the context menu
   *
   *
   * ```
   * function isItemHidden({ triggerEvent, props, data }: PredicateParams<type of props, type of data>): boolean
   * <Item hidden={isItemHidden} data={data}>content</Item>
   * ```
   */
  hidden?: BooleanPredicate;

  /**
   * Callback when the `Item` is clicked.
   *
   * @param event The event that occured on the Item node
   * @param props The props passed when you called `show(e, {props: yourProps})`
   * @param data The data defined on the `Item`
   * @param triggerEvent The event that triggered the context menu
   *
   * ```
   * function handleItemClick({ triggerEvent, event, props, data }: ItemParams<type of props, type of data>){
   *    // retrieve the id of the Item or any other dom attribute
   *    const id = e.currentTarget.id;
   *
   *    // access the props and the data
   *    console.log(props, data);
   *
   *    // access the coordinate of the mouse when the menu has been displayed
   *    const {  clientX, clientY } = triggerEvent;
   *
   * }
   *
   * <Item id="item-id" onClick={handleItemClick} data={{key: 'value'}}>Something</Item>
   * ```
   */
  onClick?: (args: ItemParams) => void;
}

export const Item: React.FC<ItemProps> = ({
  children,
  className,
  style,
  triggerEvent,
  data,
  propsFromTrigger,
  onClick = NOOP,
  disabled = false,
  hidden = false,
  ...rest
}) => {
  const refTracker = useRefTrackerContext();
  const handlerParams = {
    data,
    triggerEvent: triggerEvent as HandlerParamsEvent,
    props: propsFromTrigger,
  };
  const isDisabled = getPredicateValue(disabled, handlerParams);
  const isHidden = getPredicateValue(hidden, handlerParams);

  function handleClick(e: React.MouseEvent<HTMLElement>) {
    (handlerParams as ItemParams).event = e;
    isDisabled ? e.stopPropagation() : onClick(handlerParams as ItemParams);
  }

  function trackRef(node: HTMLElement | null) {
    if (node && !isDisabled)
      refTracker.set(node, {
        node,
        isSubmenu: false,
      });
  }

  function handleKeyDown(e: React.KeyboardEvent<HTMLElement>) {
    if (e.key === 'Enter') {
      (handlerParams as ItemParams).event = e;
      onClick(handlerParams as ItemParams);
    }
  }

  if (isHidden) return null;

  const cssClasses = cx(STYLE.item, className, {
    [`${STYLE.itemDisabled}`]: isDisabled,
  });

  return (
    <div
      {...rest}
      className={cssClasses}
      style={style}
      onClick={handleClick}
      onKeyDown={handleKeyDown}
      ref={trackRef}
      tabIndex={-1}
      role="menuitem"
      aria-disabled={isDisabled}
    >
      <div className={STYLE.itemContent}>{children}</div>
    </div>
  );
};
