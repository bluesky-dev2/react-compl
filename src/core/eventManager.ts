export type EventType = string | number | symbol;
export type Handler<T = any> = (args: T) => void;

export interface EventManager<E = EventType> {
  on<T = any>(event: E, handler: Handler<T>): EventManager<E>;
  off<T = any>(event: E, handler?: Handler<T>): EventManager<E>;
  emit<T = any>(event: E, args?: T): EventManager<E>;
}

function createEventManager<E = EventType>(): EventManager<E> {
  const eventList = new Map<E, Set<Handler>>();

  return {
    on<T = any>(event: E, handler: Handler<T>) {
      eventList.has(event)
        ? eventList.get(event)!.add(handler)
        : eventList.set(event, new Set([handler]));
      return this;
    },
    off<T = any>(event: E, handler: Handler<T>) {
      eventList.has(event) && eventList.get(event)!.delete(handler);
      return this;
    },
    emit<T = any>(event: E, args: T) {
      eventList.has(event) &&
        eventList.get(event)!.forEach((handler: Handler<T>) => {
          handler(args);
        });
      return this;
    },
  };
}

export const eventManager = createEventManager();
