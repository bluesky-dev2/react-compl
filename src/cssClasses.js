/*
* css classes mapping
* */
export default {
  MENU: 'react-contexify',
  SEPARATOR: 'react-contexify__separator',
  ITEM: 'react-contexify__item',
  ITEM_DISABLED: 'react-contexify__item--disabled',
  ITEM_DATA: 'react-contexify__item__data',
  ITEM_ICON: 'react-contexify__item__icon',
  THEME: 'react-contexify__theme--',
  ANIMATION_WILL_ENTER: 'react-contexify__will-enter--',
  ANIMATION_WILL_LEAVE: 'react-contexify__will-leave--'
};
