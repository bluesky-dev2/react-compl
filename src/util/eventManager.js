const eventManager = {
  eventList: new Map(),
  on(event, callback) {
    this.eventList.has(event) || this.eventList.set(event, new Set());
    this.eventList.get(event).add(callback);

    return () => this.eventList.get(event).delete(callback);
  },
  emit(event, ...args) {
    if (!this.eventList.has(event)) {
      console.warn(
        `<${event}> Event is not registered. Did you forgot to bind the event ?`
      );
      return false;
    }
    this.eventList.get(event).forEach(cb => cb.call(this, ...args));

    return true;
  }
};

export default eventManager;
