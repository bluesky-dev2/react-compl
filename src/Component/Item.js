import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import cssClasses from './../cssClasses';
import cx from 'classnames';

class Item extends PureComponent {
  static propTypes = {
    label: PropTypes.node.isRequired,
    icon: PropTypes.string,
    leftIcon: PropTypes.string,
    rightIcon: PropTypes.string,
    disabled: PropTypes.bool,
    onClick: PropTypes.func,
    data: PropTypes.any,
    target: PropTypes.any
  };

  static defaultProps = {
    icon: '',
    leftIcon: '',
    rightIcon: '',
    disabled: false,
    onClick: () => {}
  };

  handleClick = () => this.props.onClick(this, this.props.target);

  buildItem() {
    return (
      <div className={cssClasses.ITEM_DATA}>
        {this.hasIcon()}
        {this.props.label}
      </div>
    );
  }

  hasIcon() {
    return this.props.icon.length > 0
            ? <span className={`${cssClasses.ITEM_ICON} ${this.props.icon}`} />
            : '';
  }

  render() {
    const className = cx(cssClasses.ITEM, {
      [`${cssClasses.ITEM_DISABLED}`]: this.props.disabled
    });
    return (
      <div className={className} onClick={this.handleClick}>
        {this.buildItem()}
      </div>
    );
  }
}

export default Item;
