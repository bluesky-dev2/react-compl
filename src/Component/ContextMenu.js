/* global: window */
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import classNames from 'classnames';

import Item from './Item';
import cssClasses from './../cssClasses';
import eventManager from '../util/eventManager';
import { childrenOfType } from '../util/propValidator';

class ContextMenu extends Component {
  static propTypes = {
    id: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number
    ]).isRequired,
    children: childrenOfType(Item).isRequired,
    theme: PropTypes.string,
    animation: PropTypes.string
  };

  static defaultProps = {
    theme: null,
    animation: null
  };

  constructor(props) {
    super(props);
    this.state = {
      x: 0,
      y: 0,
      visible: false,
      target: null
    };
    this.menu = null;
  }

  componentDidMount() {
    window.addEventListener('resize', this.hide);
    eventManager.on(`display::${this.props.id}`, e => this.show(e));
  }
  
  componentWillUnmount() {
    window.removeEventListener('resize', this.hide);
    eventManager.off(`display::${this.props.id}`);
  }

  hide = () => {
    this.setState({ visible: false });
  };

  setRef = ref => {
    this.menu = ref;
  };

  setMenuPosition() {
    const browserSize = {
      width: window.innerWidth,
      height: window.innerHeight
    };

    // Get size of context
    const menuSize = {
      width: this.menu.offsetWidth,
      height: this.menu.offsetHeight
    };

    let { x, y } = this.state;

    if ((x + menuSize.width) > browserSize.width) {
      x -= ((x + menuSize.width) - browserSize.width);
    }

    if ((y + menuSize.height) > browserSize.height) {
      y -= ((y + menuSize.height) - browserSize.height);
    }

    this.setState({
      x,
      y
    });
  }

  getMousePosition(e) {

    const pos = {
      x: e.clientX,
      y: e.clientY
    };

    if (e.type === 'touchend' && (pos.x == null || pos.y == null)) {
      const touches = e.changedTouches;

      if (touches != null && touches.length > 0) {
        pos.x = touches[0].clientX;
        pos.y = touches[0].clientY;
      }
    }

    if (pos.x == null || pos.x < 0) {
      pos.x = 0;
    }

    if (pos.y == null || pos.y < 0) {
      pos.y = 0;
    }

    return pos;
  }

  getMenuItem() {
    return React.Children.map(this.props.children, child =>
      React.cloneElement(child, { target: this.state.target })
    );
  }

  getMenuStyle() {
    return {
      left: this.state.x,
      top: this.state.y,
      opacity: 1
    };
  }

  getMenuClasses() {
    return classNames(
      cssClasses.MENU,
      {
        [`react-contexify-menu__theme--${this.props.theme}`]: this.props.theme !==
        null,
        [`${this.props.animation}`]: this.props.animation !== null
      }
    );
  }

  show = e => {
    const { x, y } = this.getMousePosition(e);
    this.setState({
      visible: true,
      x,
      y,
      target: e.target
    }, this.setMenuPosition);
  };

  render() {
    return this.state.visible ? <aside
      className={cssClasses.CONTAINER}
      onClick={this.hide}
      onContextMenu={this.hide}
    >
      <div
        className={this.getMenuClasses()}
        style={this.getMenuStyle()}
        ref={this.setRef}
      >
        <div>
          {this.getMenuItem()}
        </div>
      </div>
    </aside>
      : null;
  }
}

export default ContextMenu;
