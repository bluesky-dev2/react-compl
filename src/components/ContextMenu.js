/* global: window */
import React, { Component, isValidElement } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import Item from './Item';
import styles from './styles';
import eventManager from '../util/eventManager';
import childrenOfType from '../util/childrenOfType';

class ContextMenu extends Component {
  static propTypes = {
    id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    children: childrenOfType(Item).isRequired,
    className: PropTypes.string,
    style: PropTypes.object,
    theme: PropTypes.string,
    animation: PropTypes.string
  };

  static defaultProps = {
    className: '',
    style: {},
    theme: null,
    animation: null
  };

  static THEME = {
    light: 'light',
    dark: 'dark'
  };

  static ANIMATION = {
    fade: 'fade',
    flip: 'flip',
    pop: 'pop',
    zoom: 'zoom'
  };

  state = {
    x: 0,
    y: 0,
    visible: false,
    targetNode: null
  };

  menu = null;
  refsFromProvider = null;
  token = null;
  unsub = [];

  componentDidMount() {
    this.unsub.push(
      eventManager.on(`display::${this.props.id}`, (e, refsFromProvider) =>
        this.show(e, refsFromProvider)
      )
    );

    this.unsub.push(eventManager.on('hideAll', this.hide));
  }

  componentWillUnmount() {
    this.unsub.forEach(cb => cb());
  }

  bindWindowEvent = () => {
    window.addEventListener('resize', this.hide);
    window.addEventListener('contextmenu', this.hide);
    window.addEventListener('mousedown', this.hide);
    window.addEventListener('click', this.hide);
    window.addEventListener('scroll', this.hide);
  };

  unBindWindowEvent = () => {
    window.removeEventListener('resize', this.hide);
    window.removeEventListener('contextmenu', this.hide);
    window.removeEventListener('mousedown', this.hide);
    window.removeEventListener('click', this.hide);
    window.removeEventListener('scroll', this.hide);
  };

  onMouseEnter = () => window.removeEventListener('mousedown', this.hide);

  onMouseLeave = () => window.addEventListener('mousedown', this.hide);

  hide = e => {
    // Firefox trigger a click event when you mouse up on contextmenu event
    if (
      typeof e !== 'undefined' &&
      e.button === 2 &&
      e.type !== 'contextmenu'
    ) {
      return;
    }
    this.unBindWindowEvent();
    this.setState({ visible: false });
  };

  setRef = ref => {
    this.menu = ref;
  };

  setMenuPosition() {
    const { innerWidth: windowWidth, innerHeight: windowHeight } = window;
    const { offsetWidth: menuWidth, offsetHeight: menuHeight } = this.menu;
    let { x, y } = this.state;

    if (x + menuWidth > windowWidth) {
      x -= x + menuWidth - windowWidth;
    }

    if (y + menuHeight > windowHeight) {
      y -= y + menuHeight - windowHeight;
    }

    this.setState(
      {
        x,
        y
      },
      this.bindWindowEvent
    );
  }

  getMousePosition(e) {
    const pos = {
      x: e.clientX,
      y: e.clientY
    };

    if (
      e.type === 'touchend' &&
      (pos.x === null || pos.y === null) &&
      (e.changedTouches !== null && e.changedTouches.length > 0)
    ) {
      pos.x = e.changedTouches[0].clientX;
      pos.y = e.changedTouches[0].clientY;
    }

    // just covering my ass I guess
    if (pos.x === null || pos.x < 0) {
      pos.x = 0;
    }

    if (pos.y === null || pos.y < 0) {
      pos.y = 0;
    }

    return pos;
  }

  cloneItem = item =>
    React.cloneElement(item, {
      targetNode: this.state.targetNode,
      refsFromProvider: this.refsFromProvider
    });

  getMenuItem() {
    return React.Children.map(this.props.children, this.cloneItem);
  }

  show = (e, refsFromProvider) => {
    e.stopPropagation();
    eventManager.emit('hideAll');
    this.refsFromProvider = refsFromProvider;

    const { x, y } = this.getMousePosition(e);

    this.setState(
      {
        visible: true,
        x,
        y,
        targetNode: e.target
      },
      this.setMenuPosition
    );
  };

  render() {
    const { theme, animation, style, className } = this.props;
    const cssClasses = cx(styles.menu, className, {
      [styles.theme + theme]: theme !== null,
      [styles.animationWillEnter + animation]: animation !== null
    });
    const menuStyle = {
      ...style,
      left: this.state.x,
      top: this.state.y + 1,
      opacity: 1
    };

    return this.state.visible ? (
      <div
        className={cssClasses}
        style={menuStyle}
        ref={this.setRef}
        onMouseEnter={this.onMouseEnter}
        onMouseLeave={this.onMouseLeave}
      >
        <div>{this.getMenuItem()}</div>
      </div>
    ) : null;
  }
}

export default ContextMenu;
