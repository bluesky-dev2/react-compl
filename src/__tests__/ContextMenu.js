/* eslint-env jest */
import React from 'react';
import { mount, shallow } from 'enzyme';

import ContextMenu from './../Component/ContextMenu';
import Item from './../Component/Item';
import eventManager from './../util/eventManager';
import cssClasses from './../cssClasses';

describe('ContextMenu', () => {
  it(
    'Should bind event when component did mount and unbind them when will unmount',
    () => {
      const component = mount(
        <ContextMenu id="foo">
          <Item>bar</Item>
        </ContextMenu>
      );
      expect(eventManager.eventList.has('display::foo')).toBe(true);
      expect(eventManager.eventList.has('hideAll')).toBe(true);
      component.unmount();
      expect(eventManager.eventList.has('display::foo')).toBe(false);
      expect(eventManager.eventList.has('hideAll')).toBe(false);
    });

  it('Should render null if `visible` is false', () => {
    const component = shallow(
      <ContextMenu id="foo">
        <Item>bar</Item>
      </ContextMenu>
      );
    component.setState({ visible: true });
    expect(component.html()).not.toBeNull();

    component.setState({ visible: false });
    expect(component.html()).toBeNull();
  });

  it('Should hide ContextMenu when `hide` method is called', () => {
    const component = shallow(
      <ContextMenu id="foo">
        <Item>bar</Item>
      </ContextMenu>
      );

    component.setState({ visible: true });
    component.instance().hide();

    expect(component.state('visible')).toBe(false);
    expect(component.html()).toBeNull();
  });

  it('Should merge theme and animation when defined', () => {
    const component = shallow(
      <ContextMenu id="foo" animation="foo" theme="bar">
        <Item>bar</Item>
      </ContextMenu>
      );
    component.setState({ visible: true });

    expect(component.find('.react-contexify')
        .hasClass(`${cssClasses.ANIMATION_WILL_ENTER}foo`))
        .toBe(true);

    expect(component.find('.react-contexify')
        .hasClass(`${cssClasses.THEME}bar`))
        .toBe(true);
  });
  it('Should be able to perform conditional rendering when child is different than Item', () => {
    const component = shallow(
      <ContextMenu id="foo">
        <Item>bar</Item>
        {null}
        {false}
      </ContextMenu>
    );
    component.setState({ visible: true });
    expect(component.html()).toContain('bar');
  });
});
