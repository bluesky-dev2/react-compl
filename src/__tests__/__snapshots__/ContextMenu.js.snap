// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContextMenu Should remove null Item from menu 1`] = `
<ContextMenu
  animation={null}
  className={null}
  id="foo"
  style={Object {}}
  theme={null}
>
  <div
    className="react-contexify"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "left": 0,
        "opacity": 1,
        "top": 1,
      }
    }
  >
    <div>
      <Item
        className={null}
        data={null}
        dataFromProvider={null}
        disabled={false}
        key=".$.0"
        nativeEvent={null}
        onClick={[Function]}
        refsFromProvider={null}
        style={Object {}}
      >
        <div
          className="react-contexify__item"
          onClick={[Function]}
          role="presentation"
          style={Object {}}
        >
          <div
            className="react-contexify__item__data"
          >
            bar
          </div>
        </div>
      </Item>
    </div>
  </div>
</ContextMenu>
`;
